grails-app/domain/com/unifina/domain/StreamStorageNode.groovy

```diff
@@ -1,25 +1,19 @@
 package com.unifina.domain
 
+import com.unifina.utils.IdGenerator
 import grails.persistence.Entity
 import groovy.transform.EqualsAndHashCode
 import groovy.transform.ToString
 
 @Entity
-@EqualsAndHashCode(includes = "streamId,storageNodeAddress")
+@EqualsAndHashCode(includes = "id,streamId,storageNodeAddress,dateCreated")
```

Comment:
If `id` uniquely identifies the object, we don't need to consider `streamId`, `storageNodeAddress` or `dateCreated` in `equals()` and `hashCode`?

Reply:
I think the contract for `equals()` and `hashCode()` is more complex than that.

If you are comparing two `StreamStorageNode`'s that have same `ID`, but other has been updated later objects should not be equal.

BTW I've started using https://jqno.nl/equalsverifier/ in new Streamr Java Client. I recommend that for all Java projects!

Reply:
True, if we think StreamStorageNode objects as mutable. When we just create and delete items, it would be maybe be ok think items as immutable objects. But if the objects are mutable, we could just compare `streamId` and `storageNodeAddress`?

Reply:
@teogeb Yes in theory, but we are not working with immutable objects here. If all fields of the class are defined with `final` then object is immutable and implementing `equals` and `hashCode` is pretty easy. I don't think that Hibernate allows `final` fields for object properties.

It might work with just `streamId` and `storageNodeAddress`, but there might be a corner case where `dateCreated` is needed.

If someone uses `StorageNode` (where only `id` is used in equals) as a key in a `Map` before `StorageNode` was saved to database `hashCode` is not going to work.

Since we really don't know how `StorageNode` class is going to be used it's safer to include all significant fields to `equals` and `hashCode`.

Note that again inheritance makes things super complex here: Use `instanceof` check in `equals` instead of `getClass()` comparison. This allows comparing classes that inherit each other and makes things a lot harder than necessary.